src/cgmemmgr.cpp

```diff
@@ -0,0 +1,839 @@
+    // runtime address
+    char *ptr{nullptr};
+    size_t total{0};
+    size_t avail{0};
+
+    Block(const Block&) = delete;
+    Block &operator=(const Block&) = delete;
+
+    Block() = default;
+
+    void *alloc(size_t size, size_t align)
+    {
+        size_t aligned_avail = avail & (-align);
+        if (aligned_avail < size)
+            return nullptr;
+        char *p = ptr + total - aligned_avail;
+        avail = aligned_avail - size;
+        return p;
+    }
+    void reset(void *addr, size_t size)
+    {
+        if (avail >= jl_page_size) {
+            uintptr_t end = uintptr_t(ptr) + total;
+            uintptr_t first_free = end - avail;
+            first_free = LLT_ALIGN(first_free, jl_page_size);
+            assert(first_free < end);
+            unmap_page((void*)first_free, end - first_free);
+        }
+        ptr = (char*)addr;
+        total = avail = size;
+    }
+};
+
+class RWAllocator {
+    static constexpr int nblocks = 8;
+    Block blocks[nblocks]{};
+public:
+    void *alloc(size_t size, size_t align)
+    {
+        size_t min_size = (size_t)-1;
+        int min_id = 0;
+        for (int i = 0;i < nblocks && blocks[i].ptr;i++) {
+            if (void *ptr = blocks[i].alloc(size, align))
+                return ptr;
+            if (blocks[i].avail < min_size) {
+                min_size = blocks[i].avail;
+                min_id = i;
+            }
+        }
+        size_t block_size = get_block_size(size);
+        blocks[min_id].reset(map_anon_page(block_size), block_size);
+        return blocks[min_id].alloc(size, align);
+    }
+};
+
+struct ROBlock : public Block {
+    // Possible states
+    // Allocation:
+    // * Initial allocation: `state & InitAlloc`
+    // * Followup allocation: `(state & Alloc) && !(state & InitAlloc)`
+    enum State {
+        // This block has no page protection set yet.
+        InitAlloc = (1 << 0),
+        // There is at least one allocation in this page.
+        Alloc = (1 << 1),
+        // `wr_ptr` can be directly used as write address.
+        WRInit = (1 << 2),
+        // With `WRInit` set, whether `wr_ptr` has write permission enabled.
+        WRReady = (1 << 3),
+    };
+
+    uintptr_t wr_ptr{0};
+    uint32_t state{0};
+    ROBlock() = default;
+
+    void swap(ROBlock &other)
+    {
+        std::swap(ptr, other.ptr);
+        std::swap(total, other.total);
+        std::swap(avail, other.avail);
+        std::swap(wr_ptr, other.wr_ptr);
+        std::swap(state, other.state);
+    }
+
+    ROBlock(ROBlock &&other)
+        : ROBlock()
+    {
+        swap(other);
+    }
+};
+
+struct Allocation {
+    void *wr_addr;
+    void *rt_addr;
+    size_t sz;
+    bool relocated;
+};
+
+template<bool exec>
+class ROAllocator {
+protected:
+    static constexpr int nblocks = 8;
+    ROBlock blocks[nblocks];
+    // Blocks that are done allocating but might not have all the permissions
+    // set or data copied yet.
+    SmallVector<ROBlock, 16> pendings;
```

Comment:
This could be cleared up a bit. We didn't realize that these had been removed from blocks. Also, maybe there's a better name.